File: LeetCode/236.Lowest_Common_Ancestor_of_a_Binary_Tree/Mario/236.Lowest_Common_Ancestor_of_a_Binary_Tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func lowestCommonAncestor(_ root: TreeNode?, _ p: TreeNode?, _ q: TreeNode?) -> TreeNode? {
        guard let node = root else { return nil }

        if node === q || node === p { return node }

        let right = lowestCommonAncestor(node.right, p, q)
        let left = lowestCommonAncestor(node.left, p, q)

        if right != nil, left != nil { return node }
        return left ?? right
    }
}

File: LeetCode/236.Lowest_Common_Ancestor_of_a_Binary_Tree/236.Lowest_Common_Ancestor_of_a_Binary_Tree.py
#!/usr/bin/env python3


# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        # 递归终止
        if not root or root == p or root == q:
            return root
        # 递归左右子树
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        # 情况分析
        if left and right:
            return root
        return left if left else right
